common/src/main/java/io/netty/util/NettyRuntime.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A utility class for wrapping calls to {@link Runtime}.
+ */
+public final class NettyRuntime {
+
+    private static final AtomicInteger availableProcessors = new AtomicInteger();
+
+    /**
+     * No public constructor to prevent instances from being created.
+     */
+    private NettyRuntime() {
+    }
+
+    /**
+     * Set the number of available processors.
+     *
+     * @param availableProcessors the number of available processors
+     * @throws IllegalArgumentException if the specified number of available processors is non-positive
+     * @throws IllegalStateException    if the number of available processors is already configured
+     */
+    public static void setAvailableProcessors(final int availableProcessors) {
+        if (availableProcessors <= 0) {
+            throw new IllegalArgumentException(
+                    "availableProcessors must be positive but was [" + availableProcessors + "]");
+        }
+        if (!NettyRuntime.availableProcessors.compareAndSet(0, availableProcessors)) {
+            throw new IllegalStateException(
+                    "availableProcessors is already set to [" + NettyRuntime.availableProcessors.get() + "], " +
+                            "rejecting [" + availableProcessors + "]");
+        }
+    }
+
+    /**
+     * Get the configured number of available processors. The default is {@link Runtime#availableProcessors()}. This
+     * can be overridden by setting the system property "io.netty.availableProcessors" or by invoking
+     * {@link #setAvailableProcessors(int)} before any calls to {@link #availableProcessors()}.
+     *
+     * @return the configured number of available processors
+     */
+    @SuppressForbidden(reason = "to obtain default number of available processors")
+    public static int availableProcessors() {
+        if (NettyRuntime.availableProcessors.get() == 0) {
+            final int availableProcessors =
+                    SystemPropertyUtil.getInt(
+                            "io.netty.availableProcessors",
+                            Runtime.getRuntime().availableProcessors());
+            setAvailableProcessors(availableProcessors);
```

Comment:
@jasontedor - have you considered use cases such as the following:

```java
public class Whatever {
    private final int procs;
    public Whatever() {
        procs = NettyRuntime.availableProcessors();
    }
}

public class Whatever2 {
    private final int procs;
    public Whatever2() {
        procs = NettyRuntime.availableProcessors();
    }
}

public static void main(String[] args) {
    executor.execute(() -> {
        Whatever w = new Whatever();
    });
    
    executor.execute(() -> {
        Whatever2 w = new Whatever2();
    });
// block some how to prevent exiting
}
```

It is contrived but just to demonstrate that calling the accessor may happen from two different threads and it seems unexpected for this to be considered and error if the SystemProperty initialization path races between the two threads and ultimately end up setting the same value. 

IIUC you want to differentiate the case where a `setAvailableProcessors` races with the `availableProcessors` default initialization? Maybe `setAvailableProcessors` can use something similar to `AtomicStampedReference` to differentiate from the case of `setAvailableProcessors` raced with `availableProcessors` and coincidentally set the same value as the system default ... and `availableProcessors` would only throw if it tried to set the system default but a different value existed?